rfcs/2020-04-15-wasm-foreign-modules.md

```diff
@@ -0,0 +1,559 @@
+
+We could also consider adopting a new strategy for our `event` type. However, that work is outside the scope of this
+RFC. The current structure of the `Event` type is already discussed in
+[#1891](https://github.com/timberio/vector/issues/1891).
+
+We should also consider if we want to change how we handle codecs, since it is likely that WASM module use cases will
+include wanting to add codec support to already existing sources.
+
+**Temporary Conclusion:** We will adopt hostcalls for the time being to allow for a minimal POC. We would like to
+investigate these choices lat er, and we will make a decision before stabilizing WASM modules.
+
+
+### Consider Packaging Approach
+
+We will need to expose some Vector guest API to users. This will require ABI compatibility with Vector, so it makes
+sense to keep it in the same repository (and maybe the same crate?) as Vector. We can consider either making Vector able
+to be built by WASI targets (more feature flags) or using a workspace and having the guest API as a workspace member we
+publish.
+
+A couple good questions to consider:
+
+* Do we one day want to support a cross platform `cargo install vector` option for installing a Vector binary?
+* Should a user import `vector::...` to use the Vector guest API in their Rust module?
+* Vector's internal API is largely undocumented and kind of a mess. Should we hide/clean the irrelevant stuff?
+
+**Conclusion:** We are including a workspace member `foreign_modules` in the `lib/foreign-modules` directory.
+This may be renamed and/or published in the future.
+
+
+### Consider observability
+
+How can we let users see what's happening in WASM modules? Can we use tracing somehow? Lucet supports tracing, perhaps
+we could hook in somehow?
+
+**Conclusion:** Preliminary results show we may be able to allow guests to write with the current tracing subscriber.
+
+
+### Consider Platform Support
+
+There is ongoing work to support more platforms in Lucet, here are how things stand today:
+
+
+Platform support:
+
+* [x] Linux (x86_64)
+* [ ] Linux (ARMv7) (Likely never)
+* [ ] Linux (Aarch64) (Probably coming, ARM contributing)
+* [ ] Mac (x86_64)
+  * https://github.com/bytecodealliance/lucet/pull/437
+* [ ] Windows (x86_64)
+  * https://github.com/bytecodealliance/wasmtime/pull/1216
+  * https://github.com/bytecodealliance/lucet/issues/442
+  * https://github.com/bytecodealliance/lucet/pull/437
+* [ ] FreeBSD (x86_64)
+  * https://github.com/bytecodealliance/lucet/pull/419
+  * https://github.com/bytecodealliance/lucet/pull/437
+
+**Temporary Conclusion:** The Lucet team desires to support other platforms in the near future. Since the lion's share
+of Vector usage is on X86_64 Linux, and this platform is already supported, we decided to adopt Lucet for now. We could
+consider adopting `wasmtime` in the future if lucet does not eventually reach it in terms of platform parity.
```

Comment:
Worth noting that even though most deployment targets are linux it's possible that a lot of developer environments are macos, and this means only being able to test your local configs via virtualization.

Reply:
While I don't disagree with you, I don't think "developers who code exclusively on Mac OS and deploy on Linux, but don't use a virtualization solution" is that huge a market to trade off the gains we get from lucet in performance and security.

From https://blog.rust-lang.org/2020/04/17/Rust-survey-2019.html:

![image](https://user-images.githubusercontent.com/130903/79900916-dabe7e00-83c3-11ea-96df-6610e8cd4517.png)

Reply:
Another reference point: https://insights.stackoverflow.com/survey/2019#technology-_-developers-primary-operating-systems

<img width="729" alt="Image 2020-04-21 at 2 40 01 PM" src="https://user-images.githubusercontent.com/1663286/79901685-2e3dc580-83de-11ea-8731-5c55f83fc6cb.png">